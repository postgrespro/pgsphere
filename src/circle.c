#include "circle.h"

/* Circle functions */

PG_FUNCTION_INFO_V1(spherecircle_in);
PG_FUNCTION_INFO_V1(spherecircle_equal);
PG_FUNCTION_INFO_V1(spherecircle_equal_neg);
PG_FUNCTION_INFO_V1(spherecircle_distance);
PG_FUNCTION_INFO_V1(spherecircle_point_distance);
PG_FUNCTION_INFO_V1(spherecircle_point_distance_com);
PG_FUNCTION_INFO_V1(spherepoint_in_circle);
PG_FUNCTION_INFO_V1(spherepoint_in_circle_neg);
PG_FUNCTION_INFO_V1(spherepoint_in_circle_com);
PG_FUNCTION_INFO_V1(spherepoint_in_circle_com_neg);
PG_FUNCTION_INFO_V1(spherecircle_in_circle);
PG_FUNCTION_INFO_V1(spherecircle_in_circle_neg);
PG_FUNCTION_INFO_V1(spherecircle_in_circle_com);
PG_FUNCTION_INFO_V1(spherecircle_in_circle_com_neg);
PG_FUNCTION_INFO_V1(spherecircle_overlap);
PG_FUNCTION_INFO_V1(spherecircle_overlap_neg);
PG_FUNCTION_INFO_V1(spherecircle_center);
PG_FUNCTION_INFO_V1(spherecircle_radius);
PG_FUNCTION_INFO_V1(spherepoint_to_circle);
PG_FUNCTION_INFO_V1(spherecircle_by_center);
PG_FUNCTION_INFO_V1(spherecircle_area);
PG_FUNCTION_INFO_V1(spherecircle_circ);
PG_FUNCTION_INFO_V1(spheretrans_circle);
PG_FUNCTION_INFO_V1(spheretrans_circle_inverse);


bool
scircle_eq(const SCIRCLE *c1, const SCIRCLE *c2)
{
	return (spoint_eq(&c1->center, &c2->center) &&
				FPeq(c1->radius, c2->radius));
}

bool
spoint_in_circle(const SPoint *p, const SCIRCLE *c)
{
	float8 dist = spoint_dist(p, &c->center);

	if (FPle(dist, c->radius))
	{
		return true;
	}
	return false;
}

void
euler_scircle_trans(SCIRCLE *out, const SCIRCLE *in, const SEuler *se)
{
	euler_spoint_trans(&out->center, &in->center, se);
	out->radius = in->radius;
}

Datum
spherecircle_in(PG_FUNCTION_ARGS)
{
	SCIRCLE	   *c = (SCIRCLE *) palloc(sizeof(SCIRCLE));
	char	   *s = PG_GETARG_CSTRING(0);
	double		lng,
				lat,
				radius;

	init_buffer(s);
	sphere_yyparse();
	if (get_circle(&lng, &lat, &radius))
	{
		c->center.lng = lng;
		c->center.lat = lat;
		c->radius = radius;
		reset_buffer();

		/*
		 * It's important to allow circles with radius 90 degrees!
		 */
		if (FPgt(c->radius, PIH))
		{
			pfree(c);
			c = NULL;
			elog(ERROR, "spherecircle_in: radius must be not greater than 90 degrees");
		}
		else if (FPeq(c->radius, PIH))
		{
			/* set "exact" 90 degrees */
			c->radius = PIH;
		}
		spoint_check(&c->center);
	}
	else
	{
		reset_buffer();
		pfree(c);
		c = NULL;
		elog(ERROR, "spherecircle_in: parse error");
	}
	PG_RETURN_POINTER(c);
}

Datum
spherecircle_equal(PG_FUNCTION_ARGS)
{
	SCIRCLE	   *c1 = (SCIRCLE *) PG_GETARG_POINTER(0);
	SCIRCLE	   *c2 = (SCIRCLE *) PG_GETARG_POINTER(1);

	PG_RETURN_BOOL(scircle_eq(c1, c2));
}

Datum
spherecircle_equal_neg(PG_FUNCTION_ARGS)
{
	SCIRCLE	   *c1 = (SCIRCLE *) PG_GETARG_POINTER(0);
	SCIRCLE	   *c2 = (SCIRCLE *) PG_GETARG_POINTER(1);

	PG_RETURN_BOOL(!scircle_eq(c1, c2));
}

Datum
spherecircle_distance(PG_FUNCTION_ARGS)
{
	SCIRCLE	   *c1 = (SCIRCLE *) PG_GETARG_POINTER(0);
	SCIRCLE	   *c2 = (SCIRCLE *) PG_GETARG_POINTER(1);
	float8		dist = spoint_dist(&c1->center, &c2->center);

	dist -= (c1->radius + c2->radius);
	if (dist < 0.0)
	{
		dist = 0.0;
	}
	PG_RETURN_FLOAT8(dist);
}

Datum
spherecircle_point_distance(PG_FUNCTION_ARGS)
{
	SCIRCLE	   *c = (SCIRCLE *) PG_GETARG_POINTER(0);
	SPoint	   *p = (SPoint *) PG_GETARG_POINTER(1);
	float8		dist = spoint_dist(&c->center, p);

	dist = dist - c->radius;
	if (dist < 0.0)
	{
		dist = 0.0;
	}
	PG_RETURN_FLOAT8(dist);
}

Datum
spherecircle_point_distance_com(PG_FUNCTION_ARGS)
{
	SPoint	   *p = (SPoint *) PG_GETARG_POINTER(0);
	SCIRCLE	   *c = (SCIRCLE *) PG_GETARG_POINTER(1);
	float8		dist = spoint_dist(&c->center, p);

	dist = dist - c->radius;
	if (dist < 0.0)
	{
		dist = 0.0;
	}
	PG_RETURN_FLOAT8(dist);
}

Datum
spherepoint_in_circle(PG_FUNCTION_ARGS)
{
	SPoint	   *p = (SPoint *) PG_GETARG_POINTER(0);
	SCIRCLE	   *c = (SCIRCLE *) PG_GETARG_POINTER(1);

	PG_RETURN_BOOL(spoint_in_circle(p, c));
}

Datum
spherepoint_in_circle_neg(PG_FUNCTION_ARGS)
{
	SPoint	   *p = (SPoint *) PG_GETARG_POINTER(0);
	SCIRCLE    *c = (SCIRCLE *) PG_GETARG_POINTER(1);

	PG_RETURN_BOOL(!spoint_in_circle(p, c));
}

Datum
spherepoint_in_circle_com(PG_FUNCTION_ARGS)
{
	SPoint	   *p = (SPoint *) PG_GETARG_POINTER(1);
	SCIRCLE	   *c = (SCIRCLE *) PG_GETARG_POINTER(0);

	PG_RETURN_BOOL(spoint_in_circle(p, c));
}

Datum
spherepoint_in_circle_com_neg(PG_FUNCTION_ARGS)
{
	SPoint	   *p = (SPoint *) PG_GETARG_POINTER(1);
	SCIRCLE	   *c = (SCIRCLE *) PG_GETARG_POINTER(0);

	PG_RETURN_BOOL(!spoint_in_circle(p, c));
}

Datum
spherecircle_in_circle(PG_FUNCTION_ARGS)
{
	SCIRCLE	   *c1 = (SCIRCLE *) PG_GETARG_POINTER(0);
	SCIRCLE	   *c2 = (SCIRCLE *) PG_GETARG_POINTER(1);
	float8		dist = spoint_dist(&c1->center, &c2->center);

	if (scircle_eq(c1, c2))
	{
		PG_RETURN_BOOL(true);
	}
	else if (FPle((dist + c1->radius), c2->radius))
	{
		PG_RETURN_BOOL(true);
	}
	else
	{
		PG_RETURN_BOOL(false);
	}
}

Datum
spherecircle_in_circle_neg(PG_FUNCTION_ARGS)
{
	SCIRCLE	   *c1 = (SCIRCLE *) PG_GETARG_POINTER(0);
	SCIRCLE	   *c2 = (SCIRCLE *) PG_GETARG_POINTER(1);
	float8		dist = spoint_dist(&c1->center, &c2->center);

	if (scircle_eq(c1, c2))
	{
		PG_RETURN_BOOL(false);
	}
	else if (FPgt((dist + c1->radius), c2->radius))
	{
		PG_RETURN_BOOL(true);
	}
	else
	{
		PG_RETURN_BOOL(false);
	}
}

Datum
spherecircle_in_circle_com(PG_FUNCTION_ARGS)
{
	SCIRCLE	   *c1 = (SCIRCLE *) PG_GETARG_POINTER(1);
	SCIRCLE	   *c2 = (SCIRCLE *) PG_GETARG_POINTER(0);
	float8		dist = spoint_dist(&c1->center, &c2->center);

	if (scircle_eq(c1, c2))
	{
		PG_RETURN_BOOL(true);
	}
	else if (FPle((dist + c1->radius), c2->radius))
	{
		PG_RETURN_BOOL(true);
	}
	else
	{
		PG_RETURN_BOOL(false);
	}
}

Datum
spherecircle_in_circle_com_neg(PG_FUNCTION_ARGS)
{
	SCIRCLE	   *c1 = (SCIRCLE *) PG_GETARG_POINTER(1);
	SCIRCLE	   *c2 = (SCIRCLE *) PG_GETARG_POINTER(0);
	float8		dist = spoint_dist(&c1->center, &c2->center);

	if (scircle_eq(c1, c2))
	{
		PG_RETURN_BOOL(false);
	}
	else if (FPgt((dist + c1->radius), c2->radius))
	{
		PG_RETURN_BOOL(true);
	}
	else
	{
		PG_RETURN_BOOL(false);
	}
}

Datum
spherecircle_overlap(PG_FUNCTION_ARGS)
{
	SCIRCLE	   *c1 = (SCIRCLE *) PG_GETARG_POINTER(0);
	SCIRCLE	   *c2 = (SCIRCLE *) PG_GETARG_POINTER(1);
	float8		dist = spoint_dist(&c1->center, &c2->center);

	if (scircle_eq(c1, c2))
	{
		PG_RETURN_BOOL(true);
	}
	else if (FPlt((c1->radius + c2->radius), dist))
	{
		PG_RETURN_BOOL(false);
	}
	else
	{
		PG_RETURN_BOOL(true);
	}
}

Datum
spherecircle_overlap_neg(PG_FUNCTION_ARGS)
{
	SCIRCLE	   *c1 = (SCIRCLE *) PG_GETARG_POINTER(0);
	SCIRCLE	   *c2 = (SCIRCLE *) PG_GETARG_POINTER(1);
	float8		dist = spoint_dist(&c1->center, &c2->center);

	if (scircle_eq(c1, c2))
	{
		PG_RETURN_BOOL(false);
	}
	else if (FPge((c1->radius + c2->radius), dist))
	{
		PG_RETURN_BOOL(false);
	}
	else
	{
		PG_RETURN_BOOL(true);
	}
}

Datum
spherecircle_center(PG_FUNCTION_ARGS)
{
	SCIRCLE	   *c = (SCIRCLE *) PG_GETARG_POINTER(0);
	SPoint	   *p = (SPoint *) palloc(sizeof(SPoint));

	memcpy((void *) p, (void *) &c->center, sizeof(SPoint));
	PG_RETURN_POINTER(p);
}

Datum
spherecircle_radius(PG_FUNCTION_ARGS)
{
	SCIRCLE	   *c = (SCIRCLE *) PG_GETARG_POINTER(0);

	PG_RETURN_FLOAT8(c->radius);
}

Datum
spherepoint_to_circle(PG_FUNCTION_ARGS)
{
	SPoint	   *p = (SPoint *) PG_GETARG_POINTER(0);
	SCIRCLE	   *c = (SCIRCLE *) palloc(sizeof(SCIRCLE));

	memcpy((void *) &c->center, (void *) p, sizeof(SPoint));
	c->radius = 0;
	PG_RETURN_POINTER(c);
}

Datum
spherecircle_by_center(PG_FUNCTION_ARGS)
{
	SPoint	   *p = (SPoint *) PG_GETARG_POINTER(0);
	float8		rad = PG_GETARG_FLOAT8(1);
	SCIRCLE	   *c;

	if (FPgt(rad, PIH) || FPlt(rad, 0.0))
	{
		elog(ERROR, "radius must be not greater than 90 degrees or less than 0");
		PG_RETURN_NULL();
	}
	c = (SCIRCLE *) palloc(sizeof(SCIRCLE));
	memcpy((void *) &c->center, (void *) p, sizeof(SPoint));
	c->radius = rad;
	PG_RETURN_POINTER(c);
}

Datum
spherecircle_area(PG_FUNCTION_ARGS)
{
	SCIRCLE	   *c = (SCIRCLE *) PG_GETARG_POINTER(0);

	PG_RETURN_FLOAT8(PID * (1 - cos(c->radius)));
}

Datum
spherecircle_circ(PG_FUNCTION_ARGS)
{
	SCIRCLE	   *c = (SCIRCLE *) PG_GETARG_POINTER(0);

	PG_RETURN_FLOAT8(PID * (sin(c->radius)));
}

Datum
spheretrans_circle(PG_FUNCTION_ARGS)
{
	SCIRCLE	   *sc = (SCIRCLE *) PG_GETARG_POINTER(0);
	SEuler	   *se = (SEuler *) PG_GETARG_POINTER(1);
	SCIRCLE	   *out = (SCIRCLE *) palloc(sizeof(SCIRCLE));

	euler_scircle_trans(out, sc, se);
	PG_RETURN_POINTER(out);
}

Datum
spheretrans_circle_inverse(PG_FUNCTION_ARGS)
{
	Datum		sc = PG_GETARG_DATUM(0);
	SEuler	   *se = (SEuler *) PG_GETARG_POINTER(1);
	SEuler		tmp;
	Datum		ret;

	spheretrans_inverse(&tmp, se);
	ret = DirectFunctionCall2(spheretrans_circle,
							  sc, PointerGetDatum(&tmp));
	PG_RETURN_DATUM(ret);
}
