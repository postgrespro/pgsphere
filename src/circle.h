#ifndef __PGS_CIRCLE_H__
#define __PGS_CIRCLE_H__

#include "euler.h"

/* Spherical circle declarations */

/*
 * Spherical circle data structure: center and radius.
 */
typedef struct
{
	SPoint	center;		/* the center of circle */
	float8	radius;		/* the circle radius in radians */
} SCIRCLE;

/*
 * Checks whether two circles are equal.
 */
bool	scircle_eq(const SCIRCLE *c1, const SCIRCLE *c2);

/*
 * Checks whether a circle contains a point.
 */
bool	spoint_in_circle(const SPoint *p, const SCIRCLE *c);

/*
 * Transforms a circle using an Euler transformation.
 */
void	euler_scircle_trans(SCIRCLE *out, const SCIRCLE *in, const SEuler *se);

/*
 * Takes the input and stores it as a spherical circle.
 */
Datum	spherecircle_in(PG_FUNCTION_ARGS);

/*
 * Checks whether two circles are equal.
 */
Datum	spherecircle_equal(PG_FUNCTION_ARGS);

/*
 * Checks whether two circles are not equal.
 */
Datum	spherecircle_equal_neg(PG_FUNCTION_ARGS);

/*
 * Calculate the distance of two circles. If they overlap, this function
 * returns 0.0.
 */
Datum	spherecircle_distance(PG_FUNCTION_ARGS);

/*
 * Calculate the distance of a circle and a point. If a circle contains a point,
 * this function returns 0.0.
 */
Datum	spherecircle_point_distance(PG_FUNCTION_ARGS);

/*
 * Calculate the distance of a point and a circle. If a circle contains a point,
 * this function returns 0.0.
 */
Datum	spherecircle_point_distance_com(PG_FUNCTION_ARGS);

/*
 * Checks whether a circle contains a point.
 */
Datum	spherepoint_in_circle(PG_FUNCTION_ARGS);

/*
 * Checks whether a circle doesn't contain a point.
 */
Datum	spherepoint_in_circle_neg(PG_FUNCTION_ARGS);

/*
 * Checks whether a circle contains a point.
 */
Datum	spherepoint_in_circle_com(PG_FUNCTION_ARGS);

/*
 * Checks whether a circle doesn't contain a point.
 */
Datum	spherepoint_in_circle_com_neg(PG_FUNCTION_ARGS);

/*
 * Checks whether a circle is contained by other circle.
 */
Datum	spherecircle_in_circle(PG_FUNCTION_ARGS);

/*
 * Checks whether a circle is not contained by other circle.
 */
Datum	spherecircle_in_circle_neg(PG_FUNCTION_ARGS);

/*
 * Checks whether a circle contains other circle.
 */
Datum	spherecircle_in_circle_com(PG_FUNCTION_ARGS);

/*
 * Checks whether circle does not contain other circle.
 */
Datum	spherecircle_in_circle_com_neg(PG_FUNCTION_ARGS);

/*
 * Checks whether two circles overlap.
 */
Datum	spherecircle_overlap(PG_FUNCTION_ARGS);

/*
 * Checks whether two circles overlap.
 */
Datum	spherecircle_overlap_neg(PG_FUNCTION_ARGS);

/*
 * Returns the center of a circle.
 */
Datum	spherecircle_center(PG_FUNCTION_ARGS);

/*
 * Returns the radius of a circle.
 */
Datum	spherecircle_radius(PG_FUNCTION_ARGS);

/*
 * Converts a point to a circle.
 */
Datum	spherepoint_to_circle(PG_FUNCTION_ARGS);

/*
 * Creates a circle from center and radius.
 */
Datum	spherecircle_by_center(PG_FUNCTION_ARGS);

/*
 * Calculates the area of a circle in square radians.
 */
Datum	spherecircle_area(PG_FUNCTION_ARGS);

/*
 * Calculates the circumference of a circle in radians.
 */
Datum	spherecircle_circ(PG_FUNCTION_ARGS);

/*
 * Transforms a circle using an Euler transformation.
 */
Datum	spheretrans_circle(PG_FUNCTION_ARGS);

/*
 * Inverse transformation of a circle using an Euler transformation.
 */
Datum	spheretrans_circle_inverse(PG_FUNCTION_ARGS);

#endif
