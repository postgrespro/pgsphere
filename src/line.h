#ifndef __PGS_LINE_H__
#define __PGS_LINE_H__

#include "circle.h"

/* Spherical line declarations. */

/*
 * Spherical line data structures.
 */
typedef struct
{
	float8	phi,	/* the first rotation angle around z axis */
			theta,	/* the second rotation angle around x axis */
			psi;	/* the last rotation angle around z axis */
	float8	length;	/* the length of the line */
} SLine;

/* PGS_RELATIONSHIPS Object relationships */

/* PGS_CIRCLE_LINE_REL Circle and line */
#define PGS_CIRCLE_LINE_AVOID	0	/* circle avoids line */
#define PGS_CIRCLE_CONT_LINE	1	/* circle contains line */
#define PGS_CIRCLE_LINE_OVER	2	/* circle overlaps line */

/* PGS_LINE_LINE_REL Line and line */
#define PGS_LINE_AVOID			1	/* line avoids other line */
#define PGS_LINE_EQUAL			2	/* lines are equal */
#define PGS_LINE_CONT_LINE		3	/* line contains line */
#define PGS_LINE_CROSS			4	/* lines cross each other */
#define PGS_LINE_CONNECT		5	/* line are "connected" */
#define PGS_LINE_OVER			6	/* lines overlap each other */

/*
 * Makes a line with the starting point 'pbeg' and the ending point 'pend'. Result
 * is placed into sl.
 *
 * Returns false if the distance between the 'pbeg' and the 'pend' is 180deg.
 */
bool	sline_from_points(SLine *sl, const SPoint *pbeg, const SPoint *pend);

/*
 * Returns a meridian line of a given longitude in radians. The result is placed
 * into 'sl'.
 */
void	sline_meridian(SLine *sl, float8 lng);

/*
 * Returns the starting point of a line 'l'. Result is placed into 'p'.
 */
void	sline_begin(SPoint *p, const SLine *l);

/*
 * Returns the ending point of a line 'l'. Result is placed into 'p'.
 */
void	sline_end(SPoint *p, const SLine *l);

/*
 * Puts the minimum and the maximum latitudes of a spherical line 's1' into 'minlat'
 * and 'maxlat'.
 */
void	sline_min_max_lat(const SLine *sl, float8 *minlat, float8 *maxlat);

/*
 * Calculates spherical points with a latitude 'lat' on a spherical line.
 *
 * Returns the number of found points or <0 if undefined.
 */
int32	sphereline_latitude_points(const SLine *sl, float8 lat, SPoint *p1, SPoint *p2);

/*
 * Returns true if two lines are equal.
 */
bool	sline_eq(const SLine *l1, const SLine *l2);

/*
 * Returns the relationship between a line and a circle as PGS_CIRCLE_LINE_REL
 * int8 value.
 */
int8	sphereline_circle_pos(const SLine *sl, const SCIRCLE *sc);

/*
 * Assuming that a line and a circle overlap, this function returns true
 * if the line and the circle are touching. Make sure that the line and the
 * circle overlap before calling this function! Otherwise, the result will be
 * undefined.
 *
 * See sphereline_circle_pos (const SLine *, const SCIRCLE *)
 */
bool	sline_circle_touch(const SLine *sl, const SCIRCLE *sc);

/*
 * Returns the relationship between two lines as PGS_LINE_LINE_REL int8 value.
 */
int8	sline_sline_pos(const SLine *l1, const SLine *l2);

/*
 * Checks whether a point is on a line.
 */
bool	spoint_at_sline(const SPoint *p, const SLine *sl);

/*
 * Returns the Euler transformation of a line.
 *
 * See spheretrans_from_line(PG_FUNCTION_ARGS)
 */
void	sphereline_to_euler(SEuler *se, const SLine *sl);

/*
 * Returns the inverse Euler transformation of a line.
 *
 * See spheretrans_from_line(PG_FUNCTION_ARGS)
 */
void	sphereline_to_euler_inv(SEuler *se, const SLine *sl);

/*
 * Transforms a line using an Euler transformation.
 *
 * out - pointer to the resulting line
 * in  - pointer to the original line
 * se  - pointer to the Euler transformation
 *
 * See spheretrans_line (PG_FUNCTION_ARGS)
 */
void	euler_sline_trans(SLine *out, const SLine *in, const SEuler *se);

/*
 * Puts the center of a line 'sl' into point 'c'.
 */
void	sline_center(SPoint *c, const SLine *sl);

/*
 * The input function for spherical line.
 */
Datum	sphereline_in(PG_FUNCTION_ARGS);

/*
 * Create a line from a spherical point.
 */
Datum	sphereline_from_point(PG_FUNCTION_ARGS);

/*
 * This function creates a spherical line using a starting point
 * and an ending point. The distance between the points must not be
 * equal to 180deg.
 */
Datum	sphereline_from_points(PG_FUNCTION_ARGS);

/*
 * This function creates a spherical line using a given Euler transformation
 * and the length of a line. If the length is less than zero, an error occurs.
 * If the length is larger than 360deg, it is set to 360deg.
 */
Datum	sphereline_from_trans(PG_FUNCTION_ARGS);

/*
 * This function creates a meridian running from south to north.
 * The float8 param provides the longitude in radians.
 */
Datum	sphereline_meridian(PG_FUNCTION_ARGS);

/*
 * Swaps the starting point and the ending point of a line.
 */
Datum	sphereline_swap_beg_end(PG_FUNCTION_ARGS);

/*
 * Turns the line while preserving the starting & ending points.
 */
Datum	sphereline_turn(PG_FUNCTION_ARGS);

/*
 * Returns the beginning of a line.
 */
Datum	sphereline_begin(PG_FUNCTION_ARGS);

/*
 * Returns the ending of a line.
 */
Datum	sphereline_end(PG_FUNCTION_ARGS);

/*
 * Returns the length of a line in radians.
 */
Datum	sphereline_length(PG_FUNCTION_ARGS);

/*
 * Checks whether a line contains a point.
 */
Datum	sphereline_cont_point(PG_FUNCTION_ARGS);

/*
 * Checks whether a line doesn't contain a point.
 */
Datum	sphereline_cont_point_neg(PG_FUNCTION_ARGS);

/*
 * Checks whether a line contains a point.
 */
Datum	sphereline_cont_point_com(PG_FUNCTION_ARGS);

/*
 * Checks whether a line doesn't contain a point.
 */
Datum	sphereline_cont_point_com_neg(PG_FUNCTION_ARGS);

/*
 * Checks whether a circle contains a line.
 */
Datum	spherecircle_cont_line(PG_FUNCTION_ARGS);

/*
 * Checks whether a circle doesn't contain a line.
 */
Datum	spherecircle_cont_line_neg(PG_FUNCTION_ARGS);

/*
 * Checks whether a circle contains a line.
 */
Datum	spherecircle_cont_line_com(PG_FUNCTION_ARGS);

/*
 * Checks whether a circle doesn't contain a line.
 */
Datum	spherecircle_cont_line_com_neg(PG_FUNCTION_ARGS);

/*
 * Checks whether a circle and a line overlap.
 */
Datum	sphereline_overlap_circle(PG_FUNCTION_ARGS);

/*
 * Checks whether circle and a line don't overlap.
 */
Datum	sphereline_overlap_circle_neg(PG_FUNCTION_ARGS);

/*
 * Checks whether a circle and a line overlap.
 */
Datum	sphereline_overlap_circle_com(PG_FUNCTION_ARGS);

/*
 * Checks whether circle and a line don't overlap.
 */
Datum	sphereline_overlap_circle_com_neg(PG_FUNCTION_ARGS);

/*
 * Checks whether two lines are equal.
 */
Datum	sphereline_equal(PG_FUNCTION_ARGS);

/*
 * Checks whether two lines are not equal.
 */
Datum	sphereline_equal_neg(PG_FUNCTION_ARGS);

/*
 * Checks whether two lines cross each other.
 */
Datum	sphereline_crosses(PG_FUNCTION_ARGS);

/*
 * Checks whether two lines don't cross each other.
 */
Datum	sphereline_crosses_neg(PG_FUNCTION_ARGS);

/*
 * Checks whether two lines overlap.
 */
Datum	sphereline_overlap(PG_FUNCTION_ARGS);

/*
 * Checks whether two lines are overlap.
 */
Datum	sphereline_overlap_neg(PG_FUNCTION_ARGS);

/*
 * Returns an Euler transformation. An inverse transformation with it puts
 * the line into equator beginning at (0,0) and ending at (0,length).
 */
Datum	spheretrans_from_line(PG_FUNCTION_ARGS);

/*
 * Transforms a line with an Euler transformation.
 */
Datum	spheretrans_line(PG_FUNCTION_ARGS);

/*
 * Transforms a line with an inverse Euler transformation.
 */
Datum	spheretrans_line_inverse(PG_FUNCTION_ARGS);

#endif
