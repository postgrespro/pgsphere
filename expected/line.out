\set ECHO none
 set_sphere_output_precision 
-----------------------------
 SET 8
(1 row)

-- checking spherical line operators
SELECT sline ( spoint '(0, 90d)', spoint '(0, -89d)' )   =
       sline ( spoint '(0, 90d)', spoint '(0, -89d)' )  ;
 ?column? 
----------
 t
(1 row)

SELECT sline ( spoint '(0,  90d)', spoint '(0, -89d)' )   <>
       sline ( spoint '(0, -89d)', spoint '(0,  90d)' ) ;
 ?column? 
----------
 t
(1 row)

SELECT sline ( spoint '(0,   0d)', spoint '(10d, 0d)' )   #
       sline ( spoint '(5d,  5d)', spoint '(5d, -5d)' ) ;
 ?column? 
----------
 t
(1 row)

SELECT sline ( spoint '(0,   0d)', spoint '(10d, 0d)' )   #
       sline ( spoint '(10d,  5d)', spoint '(10d, -5d)' ) ;
 ?column? 
----------
 t
(1 row)

SELECT sline ( spoint '(0,   0d)', spoint '(10d, 0d)' )   #
       sline ( spoint '(15d,  5d)', spoint '(15d, -5d)' ) ;
 ?column? 
----------
 f
(1 row)

SELECT sline ( spoint '(0,   0d)', spoint '(10d, 0d)' )   #
       sline ( spoint '(10d,  0d)', spoint '(10d, -5d)' ) ;
 ?column? 
----------
 f
(1 row)

SELECT sline ( spoint '(0,   0d)', spoint '(10d, 0d)' )   #
       sline ( spoint '(5d,  -5d)', spoint '(5d, 5d)' ) ;
 ?column? 
----------
 t
(1 row)

SELECT sline ( spoint '(0,   0d)', spoint '(10d, 0d)' )   #
       sline ( spoint '(10d, -5d)', spoint '(10d, 5d)' ) ;
 ?column? 
----------
 t
(1 row)

SELECT sline ( spoint '(0,   0d)', spoint '(10d, 0d)' )   #
       sline ( spoint '(15d, -5d)', spoint '(15d, 5d)' ) ;
 ?column? 
----------
 f
(1 row)

SELECT sline ( spoint '(0,   0d)', spoint '(10d, 0d)' )   #
       sline ( spoint '(10d,  0d)', spoint '(10d, 5d)' ) ;
 ?column? 
----------
 f
(1 row)

       
       
-- check small lines
SELECT sline ( spoint '(0,   0d)', spoint '(0.000001d, 0d)' )   #
       sline ( spoint '(0.0000005d,  0.0000005d)', spoint '(0.0000005d, -0.0000005d)' ) ;
 ?column? 
----------
 t
(1 row)

SELECT sline ( spoint '(0,   0d)', spoint '(0.000001d, 0d)' )   #
       sline ( spoint '(0.000001d,  0.0000005d)', spoint '(0.000001d, -0.0000005d)' ) ;
 ?column? 
----------
 t
(1 row)

SELECT sline ( spoint '(0,   0d)', spoint '(0.000001d, 0d)' )   #
       sline ( spoint '(0.0000015d,  0.0000005d)', spoint '(0.0000015d, -0.0000005d)' ) ;
 ?column? 
----------
 f
(1 row)

SELECT sline ( spoint '(0,   0d)', spoint '(0.000001d, 0d)' )   #
       sline ( spoint '(0.000001d,  0d)', spoint '(0.000001d, -0.0000005d)' ) ;
 ?column? 
----------
 f
(1 row)

SELECT sline ( spoint '(0,   0d)', spoint '(0.000001d, 0d)' )   #
       sline ( spoint '(0.0000005d,  -0.0000005d)', spoint '(0.0000005d, 0.0000005d)' ) ;
 ?column? 
----------
 t
(1 row)

SELECT sline ( spoint '(0,   0d)', spoint '(0.000001d, 0d)' )   #
       sline ( spoint '(0.000001d, -0.0000005d)', spoint '(0.000001d, 0.0000005d)' ) ;
 ?column? 
----------
 t
(1 row)

SELECT sline ( spoint '(0,   0d)', spoint '(0.000001d, 0d)' )   #
       sline ( spoint '(0.0000015d, -0.0000005d)', spoint '(0.0000015d, 0.0000005d)' ) ;
 ?column? 
----------
 f
(1 row)

SELECT sline ( spoint '(0,   0d)', spoint '(0.000001d, 0d)' )   #
       sline ( spoint '(0.000001d,  0d)', spoint '(0.000001d, 0.0000005d)' ) ;
 ?column? 
----------
 f
(1 row)

       
